### PR TITLE
GEODE-7502: Remove findbugs dependency from runtime

### PR DESCRIPTION
**Pull request by alb3rtobr**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [n/a] Have you written or updated unit tests to verify your changes?

- [n/a] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/build.gradle [hunk 1]
```diff
@@ -174,7 +174,8 @@ dependencies {
   compileOnly files("${System.getProperty('java.home')}/../lib/tools.jar")
 
   //Find bugs is used in multiple places in the code to suppress findbugs warnings
-  implementation('com.github.stephenc.findbugs:findbugs-annotations')
+  compile('com.github.stephenc.findbugs:findbugs-annotations')
```
Comment (pivotal-jbarrett) [inline geode-core/build.gradle]: I think you want `compileOnly` and `testCompileOnly`.

File: geode-core/src/test/resources/expected-pom.xml [hunk 2]
```diff
@@ -114,7 +114,7 @@
     <dependency>
       <groupId>com.github.stephenc.findbugs</groupId>
       <artifactId>findbugs-annotations</artifactId>
-      <scope>runtime</scope>
+      <scope>compile</scope>
```
Comment (pivotal-jbarrett) [inline geode-core/src/test/resources/expected-pom.xml]: I think after `compileOnly` it should be omitted from the pom entirely.

**Comment by alb3rtobr** [comment]
> `testCompile` already extends `compile`. And why are we moving away from `implementation`?

I have changed `compile` & `testCompile` to `compileOnly` & `testCompileOnly` as suggested by @pivotal-jbarrett . That simplifies the expected pom. I tried to use `testCompileOnly` only but the compilation does not work, so both dependencies are needed.

**Comment by alb3rtobr** [comment]
Im closing this PR, I want to investigate this issue more in detail

**Comment by alb3rtobr** [comment]
Build works locally for me, could it be a problem of CI infrastructure?

**Comment by pivotal-jbarrett** [comment]
@alb3rtobr There is something funny in how `compileOnly` dependencies get resolved when iterating through dependencies on a task. The same issue pops up on the `geode-assembly:docs` task. @rhoughton-pivot is a Gradle genius so I am sure he will have a 1 liner fix.

**Comment by pivotal-jbarrett** [comment]
@alb3rtobr
in geode-core add:
```
testCompileOnly(platform(project(':boms:geode-all-bom')))
```
and in geode-memcachd add:
```
compileOnly(platform(project(':boms:geode-all-bom')))
```

**Comment by pivotal-jbarrett** [comment]
I have something like this:
```
===================================================================
--- geode-memcached/build.gradle	(revision 518f92650190f52dd3b697004d8d95e38d2f4ff9)
+++ geode-memcached/build.gradle	(date 1574872144345)
@@ -20,11 +20,13 @@
 apply from: "${project.projectDir}/../gradle/publish-java.gradle"
 
 dependencies {
+  compileOnly(platform(project(':boms:geode-all-bom')))
+  compileOnly('com.github.stephenc.findbugs:findbugs-annotations')
+
   implementation(platform(project(':boms:geode-all-bom')))
   implementation(project(':geode-core'))
   implementation(project(':geode-logging'))
   implementation('org.apache.logging.log4j:log4j-api')
-  implementation('com.github.stephenc.findbugs:findbugs-annotations')
 
   testImplementation('org.mockito:mockito-core')
   testImplementation(project(':geode-junit'))
===================================================================
--- geode-assembly/build.gradle	(revision 518f92650190f52dd3b697004d8d95e38d2f4ff9)
+++ geode-assembly/build.gradle	(date 1574868433905)
@@ -360,7 +360,6 @@
             it.contains('mx4j') ||
 
             // misc jars, these should be removed from the lib dir
-            it.contains('findbugs-annotations') ||
             it.contains('geode-dependencies') ||
             it.contains('geode-jca') ||
             it.contains('geode-web') ||
===================================================================
--- geode-core/build.gradle	(revision 518f92650190f52dd3b697004d8d95e38d2f4ff9)
+++ geode-core/build.gradle	(date 1574872132480)
@@ -149,6 +149,7 @@
   //These bom dependencies are used to constrain the versions of the dependencies listed below
   api(platform(project(':boms:geode-all-bom')))
   compileOnly(platform(project(':boms:geode-all-bom')))
+  testCompileOnly(platform(project(':boms:geode-all-bom')))
   // As plugin configurations that do not extend from compile,
   // we must explicitly impose version constraints on these configurations.
   antlr platform(project(':boms:geode-all-bom'))
@@ -174,7 +175,8 @@
   compileOnly files("${System.getProperty('java.home')}/../lib/tools.jar")
 
   //Find bugs is used in multiple places in the code to suppress findbugs warnings
-  implementation('com.github.stephenc.findbugs:findbugs-annotations')
+  compileOnly('com.github.stephenc.findbugs:findbugs-annotations')
+  testCompileOnly('com.github.stephenc.findbugs:findbugs-annotations')
 
   //Jgroups is a core component of our membership system.
   implementation('org.jgroups:jgroups')
```

**Comment by alb3rtobr** [comment]
Geode is my first experience with gradle, so Im learning on the go :)
In my last test, I finally got `resolveDependencies` working by leaving the dependencies as:

`compileOnly('com.github.stephenc.findbugs:findbugs-annotations')
testCompile('com.github.stephenc.findbugs:findbugs-annotations')`

And the generated pom does not have the findbugs-annotations dependency. I will test you code, I prefer to use `testCompileOnly`. Thanks!

**Comment by rhoughton-pivot** [comment]
Awesome. Out of curiosity, is the 'src/main' code that uses findbugs
actually used in multiple modules?

**Comment by pivotal-jbarrett** [comment]
> Awesome. Out of curiosity, is the 'src/main' code that uses findbugs actually used in multiple modules?

Yes, 3 modules, geode-core:main. geode-core:test, geode-memcached:main.

**Pull request merged**